core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy

```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
     actualBody = OptionalBody.body('{"id": 100, "width": 100}'.bytes)
     expectedBody = OptionalBody.body('{"id": 100, "height": 100}'.bytes)
   }
+
+  @Unroll
+  def 'matching json bodies - return no mismatches - with ignore-order - when comparing lists'() {
+    given:
+    def expectedBody = OptionalBody.body(expected.bytes)
+    def actualBody = OptionalBody.body(actual.bytes)
+    matchers.addCategory('body')
+      .addRule('$', IgnoreOrderMatcher.INSTANCE)
+
+    expect:
+    matcher.matchBody(expectedBody, actualBody, true, matchers).empty
+
+    where:
+    expected                             | actual
+    '[1, 2, 3, 4]'                       | '[2, 3, 1, 4]'
+    '["a", "b", "c", "d"]'               | '["c", "a", "b", "d"]'
+    '[1, "b", 3, "d"]'                   | '["d", 1, 3, "b"]'
+    '[{"i": "a"}, {"i": 2}, {"i": "c"}]' | '[{"i": 2}, {"i": "c"}, {"i": "a"}]'
+  }
+
```

Comment:
I was able to think of a few more test cases for situations with lists in lists (which should or should not inherit the ignore-order matching).  There are probably some more corner cases to identify, but these first few seem to work.

```suggestion

def 'matching json bodies - return no mismatches - with ignore-order on nested lists'() {
    given:
    def expected = '[[1,2,3],[4,5,6]]';
    def actual = '[[6,4,5],[2,3,1]]';
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty
  }

  def 'matching json bodies - return no mismatches - with ignore-order and overriding equality'() {
    given:
    def expected = '[[1,2,3],[4,5,6]]';
    def actual = '[[6,4,5],[2,3,1]]';
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[0]', EqualsMatcher.INSTANCE)

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    !mismatches.empty
    mismatches*.mismatch = ['Expected [[1,2,3], [4,5,6]] to equal [[6,4,5], [2,3,1]] ignoring order of elements']
    mismatches*.path = ['$']
  }
```

Reply:
The second test fails miserably with 13 mismatches.  Likely from this rule:
 .addRule('$[0]', EqualsMatcher.INSTANCE)

Reply:
If I change it to:  '.addRule('$[0].*', EqualsMatcher.INSTANCE)' it generates no mismatches, which I think makes sense.  

I understand what you'd like to test, but think that extra code block we added negatively impacted this test case.

Reply:
The second test will not pass because .addRule('$[0]', EqualsMatcher.INSTANCE) gets applied recursively and generates 13 mismatches (not just the one listed).  I've modified the test to just check !mismatches.empty.